### PR TITLE
Allow cross-project references to const enums in isolatedModules when referenced project has preserveConstEnums

### LINKED ISSUE #37774: isolatedModules doesn't respect enabled preserveConstEnum option what the project might be build with
## Search Terms

isolatedModules, preserveConstEnum, const enum

## Suggestion

The compiler should respect/understand that the package/module was compiled with enabled `preserveConstEnums` compiler option.

## Use Cases

Let's say we have a project written in TypeScript, which uses const enums.

The project uses them because the main `const enum`'s advantage is inlining values instead of runtime accessing to the value.

In other hand the maintainers understand that const enums compiles in nothing in JS code, and that means that consumers can't use a const enums in their JS code (in some cases in TS code either, see above). Thus, they decided to enable `preserveConstEnums` compiler option to leave all const enums in JS code and allow consumers use them as well.

Up to this moment everything is good. But the next who appears on the stage is `isolatedModules` (say hello to `create-react-app`, webpack's transpile-only mode, `transpileModule` compiler's API and so on). This compiler option is used to detect that imported module could be transpiler without compilation into JS code and everything will work well. One of checks for this option is that you don't import const enums (error `TS2748: Cannot access ambient const enums when the '--isolatedModules' flag is provided`).

To summarize, if the project/package is compiled with enabled `preserveConstEnums`, in `isolatedModules` mode the compiler shouldn't fail with error about unavailability of using const enums, because that's not actually true.

## Examples

Quite possible we need to provide that information to the compiler somehow (by providing tsconfig.export.json or in package.json for instance), but I don't even know how it would be done.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### LINKED ISSUE #57833: VSCode does not show some typescript `isolatedModules` errors
Type: <b>Bug</b>

### Repro

1) Checkout https://github.com/bradzacher/bug-repros/tree/vscode-typescript-isolatedModules:
    - `git clone https://github.com/bradzacher/bug-repros.git`
    - `git fetch`
    - `git switch vscode-typescript-isolatedModules`
2) Open the folder in vscode
3) Open terminal and install in folder `npm i`
4) Open `package1/index.ts`
    - Observe there are no TS errors reported.
5) Open terminal and build `npm run build`
    - Observe there is a TS build error reported:
```
> tsc -p ./package1/tsconfig.json

package1/index.ts:3:21 - error TS2748: Cannot access ambient const enums when 'isolatedModules' is enabled.

3 export const test = Foo.Bar;
                      ~~~

Found 1 error in package1/index.ts:3
```

### Expected:

VSCode shows the error when viewing `package1/index.ts` in the IDE

### Actual:

VSCode shows no errors.

----

VS Code version: Code 1.87.2 (Universal) (863d2581ecda6849923a2118d93a088b0745d9d6, 2024-03-08T15:21:31.043Z)
OS version: Darwin arm64 23.4.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 Max (10 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|2, 2, 3|
|Memory (System)|64.00GB (20.85GB free)|
|Process Argv|--crash-reporter-id 4f7fbb36-3a81-40a0-a922-a2a72f77abfa|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (49)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-ndjson|adr|0.100.0
vscode-zipfs|arc|3.0.0
ast-grep-vscode|ast|0.1.3
vscode-bazel|Baz|0.9.0
soysyntaxchecker|Blo|3.30.0
vscode-coloured-status-bar-problems|bra|0.2.0
vscode-copy-filename|bra|0.1.1
npm-intellisense|chr|1.4.5
path-intellisense|chr|2.8.5
vscode-css-modules|cli|0.5.1
coder-remote|cod|0.1.35
vscode-clock|Com|0.0.1
vscode-markdownlint|Dav|0.54.0
vscode-eslint|dba|2.4.4
make-hidden|dev|4.0.4
dprint|dpr|0.16.0
gitlens|eam|14.9.0
EditorConfig|Edi|0.16.4
prettier-vscode|esb|10.1.0
vscode-github-actions|git|0.26.2
vscode-pull-request-github|Git|0.82.0
vscode-graphql|Gra|0.9.3
vscode-graphql-syntax|Gra|1.3.5
nix-ide|jno|0.3.1
statusbarerror|Joe|0.4.0
vscode-json5|mrm|1.0.0
remote-ssh|ms-|0.109.0
remote-ssh-edit|ms-|0.86.0
hexeditor|ms-|1.9.14
remote-explorer|ms-|0.4.3
vscode-js-profile-flame|ms-|1.0.8
vsliveshare|ms-|1.0.5918
vscode-paste-image|mus|1.0.4
inline-sql-syntax|quf|2.16.0
rust-analyzer|rus|0.3.1877
bazel-stack-vscode|Sta|1.9.8
rewrap|stk|1.16.3
code-spell-checker|str|3.0.1
vscode-stylelint|sty|1.3.0
even-better-toml|tam|0.19.2
shellcheck|tim|0.37.0
jest-snapshot-language-support|tle|1.1.1
pdf|tom|1.2.2
vscode-mdx|uni|1.8.2
vscode-lldb|vad|1.10.0
es-quotes|vil|0.2.6
vscode-icons|vsc|12.7.0
vscode-wakatime|Wak|24.4.0
vscode-proto3|zxh|0.5.5

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805:30301674
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:30899288
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythongtdpath:30769146
welcomedialogc:30910334
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pyreplss1:30897532
pythonmypyd1:30879173
pythoncet0:30885854
pythontbext0:30879054
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:30979542
d34g3935:30971562
fegfb526:30981948
bg6jg535:30979843
ccp2r6:30958158
dsvsc020:30976470

```

</details>

### PR DESCRIPTION
**Pull request by andrewbranch**
Partial fix for #37774 
Fixes #57833

### CODE REVIEW

**Comment by jakebailey** [comment]
With this change, does TS itself compile with isolatedModules enabled? I think this may have been one of (or the only?) blocker for enabling it...

**Comment by andrewbranch** [comment]
There are a few fixable failures, but yes.

File: src/compiler/program.ts [hunk 1]
```diff
@@ -3835,6 +3836,13 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
         return projectReferenceRedirects.get(projectReferencePath) || undefined;
     }
 
+    function getResolvedProjectReferenceByOutputFile(path: Path): ResolvedProjectReference | undefined {
```
Comment (sheetalkamat) [inline src/compiler/program.ts:3839]: Is this not `getRedirectReferenceForResolutionFromSourceOfProject` ?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -39705,7 +39705,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (getIsolatedModules(compilerOptions)) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;
-            if (constEnumDeclaration.flags & NodeFlags.Ambient && !isValidTypeOnlyAliasUseSite(node)) {
+            const redirect = host.getResolvedProjectReferenceByOutputFile(getSourceFileOfNode(constEnumDeclaration).resolvedPath);
+            if (constEnumDeclaration.flags & NodeFlags.Ambient && !isValidTypeOnlyAliasUseSite(node) && (!redirect || !shouldPreserveConstEnums(redirect.commandLine.options))) {
```
Comment (sheetalkamat) [inline src/compiler/checker.ts:39709]: Do we need a testcase for tsserver that this wont error there . (it wont be ambient so should be ok but just something to make sure works ok?)

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Comment by timocov** [comment]
> Allow cross-project references to const enums in isolatedModules when referenced project has preserveConstEnums 

> Fixes https://github.com/microsoft/TypeScript/issues/37774

Does it work in the build mode only (with project references) or it is trying to detect closest tsconfig for a given d.ts file? If it works for project references only, then it might not fully fix #37774.

**Comment by andrewbranch** [comment]
It only works with project references, so yeah, it’s a partial fix for #37774.

**Comment by jakebailey** [comment]
Will this break the case where someone can run `-p` on a project and it work? It seems like this will end up breaking without build mode because the references won't exist.

**Comment by andrewbranch** [comment]
Yeah, you’re right. If we want to avoid that inconsistency, we have to implement an emit change as discussed in #37774. Alternatively, we could just delete this error altogether and leave users to reference const enums at their own risk.

**Comment by sheetalkamat** [comment]
Whether run `--build` or `--p` if you have setup project references in config the result should be same.

**Comment by jakebailey** [comment]
I hate to make more work, but can we have a test to verify this? Would be helpful given I don't think this is a good idea if it doesn't work. Otherwise, this would be great and would enable a very similar fix for #51530.

**Comment by andrewbranch** [comment]
I didn’t notice this, but that is already what the unnittests/tsc/projectReferences.ts is testing.

**Comment by jakebailey** [comment]
Oh dang, you're right. Nice.

**Comment by jakebailey** [comment]
In the process of attempting to build #51530, I realized that `getRedirectReferenceForResolutionFromSourceOfProject` doesn't give you back "yourself", so you'll still get the error for something defined locally. Is that intended?

Maybe that's ok for this particular case, I'm just having trouble getting a version of this check that's "another project has the flag enabled and redirect _or_ this project has it enabled and it's not just some random file but one that will be emitted)...

**Comment by andrewbranch** [comment]
I didn’t think about that case, but I think it’s ok. Including your project’s outputs as your project’s inputs is a misconfiguration in all cases, I think.

**Comment by jakebailey** [comment]
> Including your project’s outputs as your project’s inputs is a misconfiguration in all cases, I think.

I didn't really mean including the outputs, but rather that there's no redirect for a local project by definition, and therefore the check comes out to "false" (when I am personally looking for a way to make it true for the local project files too).

**Comment by andrewbranch** [comment]
I guess I don’t understand the scenario you’re saying then—if you reference a const enum in your own project, it won’t be ambient since you’ll include the input file, and would never get this error. If you’re referencing another project, you’re likely to be referencing an output file, which makes the declaration ambient.

**Comment by jakebailey** [comment]
So the part where it won't be ambient answers my question in the sense of "ok for this particular case".

So, past that, I'm just looking ahead to #51530 where I'm trying to not error on uses of const enums when `preserveConstEnums` is true, but only if we know the value is actually going to exist (i.e. is provided by a file in the current project or referenced project where `preserveConstEnums` would have emitted the runtime enum).

**Comment by andrewbranch** [comment]
I think the same ambient/non-ambient breakdown will apply there.

**Review by jakebailey — APPROVED** [review]
Anyway that doesn't block anything; I assumed Sheetal's approval was enough but here's another one 😄

**Comment by jakebailey** [comment]
I will send a WIP PR showing what I mean for the above thing later.

**Pull request merged**